### PR TITLE
sequoia-wot: init at 0.11.0

### PR DESCRIPTION
**Pull request by Cryolitia**
## Description of changes

https://gitlab.com/sequoia-pgp/sequoia-wot

A Rust library for authenticating bindings between User IDs and certificates using OpenPGP's web of trust.  This library is designed around OpenPGP's data structures, but it does not require OpenPGP data.  Instead, it is possible to manually describe a web of trust.

This crate also includes a CLI tool, sq-wot, for authenticating bindings and exploring a web of trust.

An official introduce article here: https://sequoia-pgp.org/blog/2023/03/29/202303-pretty-graphics-for-the-web-of-trust/

Thanks for review!

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/se/sequoia-wot/package.nix
Comment (doronbehar) [inline pkgs/by-name/se/sequoia-wot/package.nix:24]: Why are these needed?

File: pkgs/by-name/se/sequoia-wot/package.nix
Comment (doronbehar) [inline pkgs/by-name/se/sequoia-wot/package.nix:43]: Did you try setting `GNUPGHOME` in a `preCheck` hook?

Comment (Cryolitia) [reply pkgs/by-name/se/sequoia-wot/package.nix:43]: I set

```nix
  preCheck = ''
    export GNUPGHOME=$(mktemp -d)
    export RUST_BACKTRACE=1
  '';
```

and failed with

```log
sequoia-wot> ---- gpg_trust_roots stdout ----
sequoia-wot> thread 'gpg_trust_roots' panicked at tests/gpg.rs:25:30:
sequoia-wot> called `Result::unwrap()` on an `Err` value: No such file or directory (os error 2)
sequoia-wot> Stack backtrace:
sequoia-wot>    0: anyhow::error::<impl core::convert::From<E> for anyhow::Error>::from
sequoia-wot>    1: gpg::gpg::export
sequoia-wot>    2: core::ops::function::FnOnce::call_once
sequoia-wot>    3: test::__rust_begin_short_backtrace
sequoia-wot>    4: test::types::RunnableTest::run
sequoia-wot>    5: test::run_test_in_process
sequoia-wot>    6: std::sys_common::backtrace::__rust_begin_short_backtrace
sequoia-wot>    7: core::ops::function::FnOnce::call_once{{vtable.shim}}
sequoia-wot>    8: std::sys::pal::unix::thread::Thread::new::thread_start
sequoia-wot>    9: start_thread
sequoia-wot>   10: __GI___clone3
sequoia-wot> stack backtrace:
sequoia-wot>    0: rust_begin_unwind
sequoia-wot>    1: core::panicking::panic_fmt
sequoia-wot>    2: core::result::unwrap_failed
sequoia-wot>    3: core::ops::function::FnOnce::call_once
sequoia-wot> note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
sequoia-wot> failures:
sequoia-wot>     gpg_trust_roots
sequoia-wot> test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
```

File: pkgs/by-name/se/sequoia-wot/package.nix
Comment (Cryolitia) [reply pkgs/by-name/se/sequoia-wot/package.nix:24]: Because building the cli requires additional features `sequoia-openpgp/default`, this feature requires a series of packages that do not exist in `Cargo.lock`. Therefore, `Cargo.lock` needs to be patched to introduce additional packages so that nix can add them to the vendor. And if `Cargo.toml` is not patched, it will fail because the package is out of date and needs to be updated.

Frankly, I have no idea what the best practices are for handling this situation, so if you could enlighten me I'd be all ears.

File: pkgs/by-name/se/sequoia-wot/package.nix
Comment (doronbehar) [inline pkgs/by-name/se/sequoia-wot/package.nix:31]:
```suggestion
    rustPlatform.bindgenHook
    installShellFiles
```

Locally I noticed these log messages during the build:

```
writing man page to /build/source/target/x86_64-unknown-linux-gnu/release/build/squoia-wot-fb11ecd97b1459ce/out/sq-wot.1
writing man page to /build/source/target/x86_64-unknown-linux-gnu/release/build/squoia-wot-fb11ecd97b1459ce/out/sq-wot-authenticate.1
writing man page to /build/source/target/x86_64-unknown-linux-gnu/release/build/squoia-wot-fb11ecd97b1459ce/out/sq-wot-lookup.1
writing man page to /build/source/target/x86_64-unknown-linux-gnu/release/build/squoia-wot-fb11ecd97b1459ce/out/sq-wot-identify.1
writing man page to /build/source/target/x86_64-unknown-linux-gnu/release/build/squoia-wot-fb11ecd97b1459ce/out/sq-wot-list.1
writing man page to /build/source/target/x86_64-unknown-linux-gnu/release/build/squoia-wot-fb11ecd97b1459ce/out/sq-wot-path.1
completion file is generated: "/build/source/target/x86_64-unknown-linux-gnu/relese/build/sequoia-wot-fb11ecd97b1459ce/out/sq-wot.bash"
completion file is generated: "/build/source/target/x86_64-unknown-linux-gnu/relese/build/sequoia-wot-fb11ecd97b1459ce/out/sq-wot.fish"
completion file is generated: "/build/source/target/x86_64-unknown-linux-gnu/relese/build/sequoia-wot-fb11ecd97b1459ce/out/_sq-wot"
completion file is generated: "/build/source/target/x86_64-unknown-linux-gnu/relese/build/sequoia-wot-fb11ecd97b1459ce/out/_sq-wot.ps1"
completion file is generated: "/build/source/target/x86_64-unknown-linux-gnu/relese/build/sequoia-wot-fb11ecd97b1459ce/out/sq-wot.elv"
```

Could you find a way to detect the paths to these files and install these? Maybe something like this (used in `sequoia-sq`) will work:

```nix
  env.ASSET_OUT_DIR = "/tmp/";

  postInstall = ''
    installManPage /tmp/man-pages/*.*
    installShellCompletion \
      --cmd sq \
      --bash /tmp/shell-completions/sq.bash \
      --fish /tmp/shell-completions/sq.fish \
      --zsh /tmp/shell-completions/_sq
  '';
```

**Review by doronbehar — DISMISSED** [review]

File: pkgs/by-name/se/sequoia-wot/package.nix
Comment (Cryolitia) [reply pkgs/by-name/se/sequoia-wot/package.nix:31]: The output path like `build/squoia-wot-fb11ecd97b1459ce/out` is random. To search these outputs in `postInstall` may be very dirty.

There is an issue about gettin the output path: https://github.com/rust-lang/cargo/issues/9661

I also opened an issue for asking upstream to support generate these files runtime: https://gitlab.com/sequoia-pgp/sequoia-wot/-/issues/56

Comment (Cryolitia) [reply pkgs/by-name/se/sequoia-wot/package.nix:31]: The output path in sequoia-wot is `OUT_DIR` in https://gitlab.com/sequoia-pgp/sequoia-wot/-/blob/v0.11.0/build.rs?ref_type=tags#L99 , not `ASSET_OUT_DIR`. It's a cargo defined variable.

Comment (doronbehar) [reply pkgs/by-name/se/sequoia-wot/package.nix:32]: > The output path in sequoia-wot is `OUT_DIR` in [gitlab.com/sequoia-pgp/sequoia-wot/-/blob/v0.11.0/build.rs?ref_type=tags#L99](https://gitlab.com/sequoia-pgp/sequoia-wot/-/blob/v0.11.0/build.rs?ref_type=tags#L99) , not `ASSET_OUT_DIR`. It's a cargo defined variable.

So did you manage to use this env var to find the man pages and shell completion files?

Comment (Cryolitia) [reply pkgs/by-name/se/sequoia-wot/package.nix:32]: > > The output path in sequoia-wot is `OUT_DIR` in [gitlab.com/sequoia-pgp/sequoia-wot/-/blob/v0.11.0/build.rs?ref_type=tags#L99](https://gitlab.com/sequoia-pgp/sequoia-wot/-/blob/v0.11.0/build.rs?ref_type=tags#L99) , not `ASSET_OUT_DIR`. It's a cargo defined variable.
> 
> So did you manage to use this env var to find the man pages and shell completion files?

cargo only set it during building, we can't get it. Upstream https://github.com/rust-lang/cargo/issues/9661 has discussed it a lot. It's now not a reasonable solution for me.

File: pkgs/by-name/se/sequoia-wot/package.nix
Comment (Cryolitia) [reply pkgs/by-name/se/sequoia-wot/package.nix:46]: updated ^_^

File: pkgs/by-name/se/sequoia-wot/package.nix
Comment (doronbehar) [reply pkgs/by-name/se/sequoia-wot/package.nix:32]: > > So did you manage to use this env var to find the man pages and shell completion files?
> 
> cargo only set it during building, we can't get it. Upstream [rust-lang/cargo#9661](https://github.com/rust-lang/cargo/issues/9661) has discussed it a lot. It's now not a reasonable solution for me.

Oh I understand. I thought you meant to say that upstream's `build.rs` file indicates we can use `$OUT_DIR` instead of `$ASSET_OUT_DIR`. In any case, all of the `sequoia-*` packages are influenced by similar issues, and the way we install these shell files is inconsistent through out all of them. I tested the following to work:

```diff
diff --git i/pkgs/by-name/se/sequoia-wot/package.nix w/pkgs/by-name/se/sequoia-wot/package.nix
index dd12f990049a..4b5c6e27ea8e 100644
--- i/pkgs/by-name/se/sequoia-wot/package.nix
+++ w/pkgs/by-name/se/sequoia-wot/package.nix
@@ -2,6 +2,7 @@
   lib,
   rustPlatform,
   fetchFromGitLab,
+  installShellFiles,
   libclang,
   pkg-config,
   nettle,
@@ -30,6 +31,7 @@ rustPlatform.buildRustPackage rec {
   nativeBuildInputs = [
     pkg-config
     rustPlatform.bindgenHook
+    installShellFiles
   ];
 
   buildInputs = [
@@ -38,10 +40,31 @@ rustPlatform.buildRustPackage rec {
     openssl
     sqlite
   ];
+  # Install shell completion files and manual pages. Unfortunatly it is hard to
+  # predict the paths to all of these files generated during the build, and it
+  # is impossible to control these using `$OUT_DIR` or alike, as implied by
+  # upstream's `build.rs`. This is a general Rust issue also discussed in
+  # https://github.com/rust-lang/cargo/issues/9661, also discussed upstream at:
+  # https://gitlab.com/sequoia-pgp/sequoia-wot/-/issues/56
+  postInstall = ''
+    installShellCompletion --cmd sq-wot \
+      --bash target/*/release/build/sequoia-wot-*/out/sq-wot.bash \
+      --fish target/*/release/build/sequoia-wot-*/out/sq-wot.fish \
+      --zsh  target/*/release/build/sequoia-wot-*/out/_sq-wot
+    # Also elv and powershell are generated there
+    installManPage \
+      target/*/release/build/sequoia-wot-*/out/sq-wot.1 \
+      target/*/release/build/sequoia-wot-*/out/sq-wot-authenticate.1 \
+      target/*/release/build/sequoia-wot-*/out/sq-wot-lookup.1 \
+      target/*/release/build/sequoia-wot-*/out/sq-wot-identify.1 \
+      target/*/release/build/sequoia-wot-*/out/sq-wot-list.1 \
+      target/*/release/build/sequoia-wot-*/out/sq-wot-path.1
+  '';
 
   buildFeatures = [ "sequoia-openpgp/default" ];
 
   nativeCheckInputs = [ gnupg ];
+  doCheck = true;
 
   meta = with lib; {
     description = "A Rust CLI tool for authenticating bindings and exploring a web of trust";
```

Which is based upon what's in `sequoia-sqop`.

Comment (Cryolitia) [reply pkgs/by-name/se/sequoia-wot/package.nix:34]: done

File: pkgs/by-name/se/sequoia-wot/package.nix
Comment (doronbehar) [reply pkgs/by-name/se/sequoia-wot/package.nix:26]: > Because building the cli requires additional features `sequoia-openpgp/default`, this feature requires a series of packages that do not exist in `Cargo.lock`. Therefore, `Cargo.lock` needs to be patched to introduce additional packages so that nix can add them to the vendor. And if `Cargo.toml` is not patched, it will fail because the package is out of date and needs to be updated.
> 
> Frankly, I have no idea what the best practices are for handling this situation, so if you could enlighten me I'd be all ears.

@Cryolitia I'm a bit bothered by these, because I worry it will create a maintenance headache in future releases. Could we maybe write an issue to upstream about this? I'm not sure I will be able to do it myself and explain it best for their point of view. Other then that, the PR looks ready. Let's give them a few days to respond and we'll decide afterwards.

Comment (Cryolitia) [reply pkgs/by-name/se/sequoia-wot/package.nix:26]: > @Cryolitia I'm a bit bothered by these, because I worry it will create a maintenance headache in future releases. Could we maybe write an issue to upstream about this? I'm not sure I will be able to do it myself and explain it best for their point of view. Other then that, the PR looks ready. Let's give them a few days to respond and we'll decide afterwards.

I think it's more that a nixpkgs's problem, we should figure out a way to vendor all build-time dependency packages, not only default in `cargo.lock`. I wonder if it not happed before in nixpkgs, I think it should be more common.

Comment (doronbehar) [reply pkgs/by-name/se/sequoia-wot/package.nix:26]: I have a simple question regarding upstream: If they had included your 2 patches, would they be affected when they'd build their project locally or on their CI? What is the reason they don't enable all features or all the default features of `sequoia-openpgp/default`?

Comment (doronbehar) [reply pkgs/by-name/se/sequoia-wot/package.nix:26]: > it is true that the default backend should not be set and users should be reminded to make their own choices.

Maybe my problem is that I don't understand this sentence - what is the default backend you are talking about here?

Comment (Cryolitia) [reply pkgs/by-name/se/sequoia-wot/package.nix:26]: You may have noticed that there's one line in `packages.nix`: `buildFeatures = [ "sequoia-openpgp/default" ];`
Without this feature, the build would failed with error below:
```log
sequoia-wot> error: failed to run custom build command for `sequoia-openpgp v1.19.0`
sequoia-wot> Caused by:
sequoia-wot>   process didn't exit successfully: `/build/source/target/release/build/sequoia-openpgp-81b84cfda6136cb6/build-script-build` (exit status: 1)
sequoia-wot>   --- stderr
sequoia-wot>   No cryptographic backend selected.
sequoia-wot>   Sequoia requires a cryptographic backend.  This backend is selected at compile
sequoia-wot>   time using feature flags.
sequoia-wot>   See https://crates.io/crates/sequoia-openpgp#crypto-backends
sequoia-wot> warning: build failed, waiting for other jobs to finish...
```
That means that we are required to choose a cryptographic backend manually and explicitly, they don't provide a defalt backend.

In their [README](https://gitlab.com/sequoia-pgp/sequoia/-/tree/main/openpgp?ref_type=heads#crypto-backends) is documented that, the `openpgp` provides some cryptographic backend, the `default` backend is nettle. Notice that the `default` is just named default, it's not truly the default backend the should be enabled by default. 

It is obvious that, the `sequoia-wot` deliberately set the feature `openpgp/default` backend not enabled by default, but force user —— or package maintainer to choose a backend. So I guess they won't accept the patch.

Comment (Cryolitia) [reply pkgs/by-name/se/sequoia-wot/package.nix:26]: The problem is that, the crates in the patch such as `bzip2` are just required by the feature `sequoia-openpgp/default`, which means a default backend `nettle` named `default`, but may not for other `sequoia-openpgp`'s backend.

Comment (doronbehar) [reply pkgs/by-name/se/sequoia-wot/package.nix]: @Cryolitia thanks for explaining, I pushed a commit that changes this, tell me what do you think about it.

Comment (Cryolitia) [reply pkgs/by-name/se/sequoia-wot/package.nix]: Done. You have great insights into Rust, appreciate it.

**Comment by doronbehar** [comment]
I saw this error on ofborg's Darwin build attempt:

```
tmp/nix-build-sequoia-wot-0.11.0.drv-0/source/target/x86_64-apple-darwin/release/deps/sq_wot_audit_data.o" "-Wl,-u,_AUDITABLE_VERSION_INFO"
  = note: ld: framework not found SystemConfiguration
          clang-16: error: linker command failed with exit code 1 (use -v to see invocation)
```

So I pushed an attempt to fix it.

**Comment by doronbehar** [comment]
I also removed `libclang` from the `buildInputs`, because it is not really required if you also include `rustPlatform.bingenHook`.

**Comment by doronbehar** [comment]
[`nix-init`](https://github.com/nix-community/nix-init) is pretty good at guessing `buildInputs` of rust packages apparently.

**Comment by Cryolitia** [comment]
Should I sign the commit and re-push? Or I could just waiting, doing nothing needed.

**Comment by doronbehar** [comment]
> Should I sign the commit and re-push? Or I could just waiting, doing nothing needed.

I didn't mean to ask you to do anything, I only shared what I did, because it's hard to tell the difference with all of those force pushes. As for the signing, it is up to you - for Nixpkgs and for me it doesn't matter if you sign it or not. Force pushing now again will disrupt CI once more so consider that as well.

**Comment by Cryolitia** [comment]
Ok thx

**Comment by Cryolitia** [comment]
Ok thx

**Review by nicoonoclaste — APPROVED** [review]
Looks good, OK to merge as far as I'm concerned, which I'll happily do once my question is answered

File: pkgs/by-name/se/sequoia-wot/package.nix
Comment (doronbehar) [reply pkgs/by-name/se/sequoia-wot/package.nix:56]: > Does this mean `buildInputs` shouldn't include `nettle` on Windows?

It is a bit perfectionist, because I doubt ever Windows will be a system people will actually use, but you are correct :) Thanks.

**Comment by doronbehar** [comment]
CI was completely green before, so I won't wait for darwin CI.

**Comment by doronbehar** [comment]
Thanks @Cryolitia for contributing this package :+1: .

**Pull request merged**
